python/ray/train/huggingface/accelerate/_accelerate_utils.py

```diff
@@ -0,0 +1,172 @@
+from contextlib import nullcontext
+from pathlib import Path
+
+try:
+    from packaging.version import Version
+except ImportError:
+    from distutils.version import LooseVersion as Version
+
+import accelerate
+
+if Version(accelerate.__version__) < Version("0.17.0.dev0"):
+    raise ImportError(
+        f"AccelerateTrainer requires accelerate>=0.17.0, got {accelerate.__version__}"
+    )
+
+from accelerate.commands.launch import (
+    ComputeEnvironment,
+    _validate_launch_command,
+    launch_command_parser,
+    prepare_deepspeed_cmd_env,
+    prepare_multi_gpu_env,
+    prepare_simple_launcher_cmd_env,
+)
+from accelerate.utils import is_deepspeed_available
+from accelerate.commands.config import default_config_file, load_config_from_file
+from accelerate.commands.config.config_args import ClusterConfig
+
+logger = logging.getLogger(__name__)
+
+
+class AccelerateDefaultNamespace(Namespace):
+    @property
+    def parser(self):
+        return launch_command_parser()
+
+    def __getattr__(self, name: str):
+        if name == "training_script_args":
+            return []
+        return self.parser.get_default(name)
+
+
+class AccelerateConfigWrapper:
+    """
+    Lets Trainables know to treat this as already loaded file content instead of path.
+    """
+
+    def __init__(
+        self, config_raw: str, deepspeed_config_raw: Optional[str] = None
+    ) -> None:
+        self.config_raw = config_raw
+        self.deepspeed_config_raw = deepspeed_config_raw
+
+    def __bool__(self) -> bool:
+        return bool(self.config_raw)
+
+
+def simple_launcher(args):
+    _, current_env = prepare_simple_launcher_cmd_env(args)
+
+    os.environ.update(current_env)
+
+
+def multi_gpu_launcher(args):
+    current_env = prepare_multi_gpu_env(args)
+    os.environ.update(current_env)
+
+
+def deepspeed_launcher(args):
+    if not is_deepspeed_available():
+        raise ImportError(
+            "DeepSpeed is not installed => run `pip3 install deepspeed` "
+            "or build it from source."
+        )
+
+    _, current_env = prepare_deepspeed_cmd_env(args)
+
+    os.environ.update(current_env)
+
+
+def launch_command(args):
+    args, defaults, mp_from_config_flag = _validate_launch_command(args)
+
+    # Use the proper launcher
+    if args.use_deepspeed and not args.cpu:
+        args.deepspeed_fields_from_accelerate_config = (
+            list(defaults.deepspeed_config.keys()) if defaults else []
+        )
+        if mp_from_config_flag:
+            args.deepspeed_fields_from_accelerate_config.append("mixed_precision")
+        args.deepspeed_fields_from_accelerate_config = ",".join(
+            args.deepspeed_fields_from_accelerate_config
+        )
+        deepspeed_launcher(args)
+    elif args.use_fsdp and not args.cpu:
+        multi_gpu_launcher(args)
+    elif args.use_megatron_lm and not args.cpu:
+        multi_gpu_launcher(args)
+    elif args.multi_gpu and not args.cpu:
+        multi_gpu_launcher(args)
```

Comment:
huh, these are actually just:
```
if not args.cpu and (
    args.use_fsdp or args.use_megatron_lm or args.multi_gpu
):
    multi_gpu_launcher(args)
```

Reply:
this is the same structure as in accelerate itself (makes it easier to compare the code between the two)

Reply:
ok